product.py

```diff
@@ -534,13 +534,31 @@ def findXenSourceBackups():
         b = None
         try:
             b = util.TempMount(p, 'backup-', ['ro'], 'ext3')
-            if os.path.exists(os.path.join(b.mount_point, '.xen-backup-partition')):
-                backup = XenServerBackup(p, b.mount_point)
-                logger.log("Found a backup: %s" % (repr(backup),))
-                if backup.version >= XENSERVER_MIN_VERSION and \
-                        backup.version <= THIS_PLATFORM_VERSION:
-                    backups.append(backup)
-        except:
+            if not os.path.exists(os.path.join(b.mount_point, '.xen-backup-partition')):
+                raise StopIteration()
+
+            backup = XenServerBackup(p, b.mount_point)
+            logger.log("Found a backup: %s" % (repr(backup),))
+
+            if backup.version < XENSERVER_MIN_VERSION:
+                logger.log("findXenSourceBackups: ignoring, platform too old: %s < %s" %
+                           (backup.version, XENSERVER_MIN_VERSION))
+                raise StopIteration()
+            if backup.version > THIS_PLATFORM_VERSION:
+                logger.log("findXenSourceBackups: ignoring later platform: %s > %s" %
+                           (backup.version, THIS_PLATFORM_VERSION))
+                raise StopIteration()
+            if not os.path.exists(backup.root_disk):
+                logger.error("findXenSourceBackups: PRIMARY_DISK=%r does not exist" %
+                             (backup.root_disk,))
+                raise StopIteration()
+
+            backups.append(backup)
+
+        except StopIteration:
+            pass
+        except Exception as ex:
+            logger.log("findXenSourceBackups caught exception for partition %s: %s" % (p, ex))
```

Comment:
We probably want to use `logger.logException(ex)` here rather than trying to include the exception directly in the log message, as that will then give us a useful traceback etc, so e.g.:
```
logger.log("findXenSourceBackups caught exception for partition %s:" % (p,))
logger.logException(ex)
```

Reply:
fixed